### DISCUSSION TITLE
How to communicate between the different edgex services internally using Message Bus and Redis database

### DISCUSSION THREAD

**Discussion by Devendra-7** [edgexfoundry/community#232] [Q&A]
Hi All,
We are working  imx6 based board for one of our IoT project and using yocto(3.4 version) based linux images.
Successfully we have integrated few edgex micro services (version 3.1 i.e Napa)like coredata, core-command,device-modbus,device-mqtt,device-snmp and redis database in yocto BSP, able to see the above services in the rootfs after booting the board and redis databae CLI commands are also available.
currently working on enabling the message bus support in yocto BSP.

  I have few questions regarding the database and message bus

1. If Anybody integrated the message bus(go-mod-messaging) in the yocto plsease provide the steps.  
2. core data and message bus use the same database ?.
3.  The way we Enabled/ installed  the database into the BSP is correct ? if not pls suggest the correct method to enable it .
4. How to configure the core data and message bus services to use the redis database and exchange the data between them, please provide the sample configuration for the same.

Thanks in Advance ...!

**Comment by jiekechoo** [comment]
@Devendra-7 , it's a big challenge for you. Did you run EdgeX in native linux? or try to use Docker ?

**Reply by Devendra-7** [reply]
Hi jiekechoo,
Thanks for the reply. Unfortunately i never tried running of Edgex in native linux  or Docker environment. 
It would be great if you share the configurations for the core-data and message bus to use the redis database for native linux or docker environment. I am working with Edgex Napa version.

**Reply by jiekechoo** [reply]
What's your imx6 host configuration? like: CPU arch and cores, RAM, HARD DRIVER, ...

And you can build and run all EdgeX services in native Linux kernel, if you have installed Golang env. but it's unusual.

> If Anybody integrated the message bus(go-mod-messaging) in the yocto plsease provide the steps.

`go-mod-messaging ` is a library, you don't deploy it. If you want to develop device service or app service, just refer the getting-started article: https://docs.edgexfoundry.org/3.1/getting-started/Ch-GettingStartedSDK/ .

> core data and message bus use the same database ?.

Core Data is completely optional. You can refer this article: https://docs.edgexfoundry.org/3.1/microservices/core/data/Purpose/ . And you could change the message bus, see this: https://docs.edgexfoundry.org/3.1/microservices/general/messagebus/ . By default, Core Data and Message Bus are using Redis.

> The way we Enabled/ installed the database into the BSP is correct ? if not pls suggest the correct method to enable it .

You just need to install the database(like redis).

> How to configure the core data and message bus services to use the redis database and exchange the data between them, please provide the sample configuration for the same.

The Core Data Service is using the file `configuration.yaml`, here are some articles, https://docs.edgexfoundry.org/3.1/microservices/configuration/CommonConfiguration/ , https://docs.edgexfoundry.org/3.1/design/ucr/Common%20Configuration/

By the way, I suggest you to read the documentation website: docs.edgexfoundry.org , and you can find more details from Source Code repos: github.com/edgexfoundry

**Reply by Devendra-7** [reply]
Hi jiekechoo,

Thanks for your inputs, i will  check the documents you referred.
Currently we are using imx6 based on  Arm-CortexA9 Quad core CPU  @up to 1.2GHz/core.
Memory configuration 1GB DDR3 RAM, 8GB eMMC. we have flexibility to increase the memory configuration.

Thanks in Advance ...!

**Comment by cloudxxx8** [comment]
By default, Core Data stores data to RedisDB, and EdgeX use Redis pub/sub (the same RedisDB service) as the message bus.
You should not need any additional configuration.

**Reply by M0hanrajp** [reply]
> By default, Core Data stores data to RedisDB, and EdgeX use Redis pub/sub (the same RedisDB service) as the message bus. You should not need any additional configuration.

Hello @cloudxxx8, do you know anything regarding Redis cache ? if there is any service in edgex that implements/uses redis cache ?

Thanks

**Reply by cloudxxx8** [reply]
no, I don't know about it
